hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java

```diff
@@ -5000,19 +4999,38 @@ private Predicate consumeConjunctTreatTypeRestrictions(Map<TableGroup, Map<Strin
 					tableGroup,
 					this
 			);
+			// We need to check if this is a treated left or full join, which case we should
+			// allow null discriminator values to maintain correct semantics
+			final TableGroupJoin join = getParentTableGroupJoin( tableGroup );
+			final boolean allowNulls = join != null && ( join.getJoinType() == SqlAstJoinType.LEFT || join.getJoinType() == SqlAstJoinType.FULL );
```

Comment:
Updated the solution to allow nulless in the discriminator predicate (i.e. added `or t.disc_col is null`) when the treated table group is a `LEFT` or `FULL` join.